// wcs-core/dist/collection/components/com-nav-item/com-nav-item.js
import { Host, h } from "@stencil/core";
/**
 * The com-nav-item is a subcomponent of `wcs-com-nav`. It represents a list-item wrapper around a link.
 *
 * @cssprop --wcs-com-nav-item-gap - Gap between text and arrow
 * @cssprop --wcs-com-nav-item-font-weight - Font-weight of item
 * @cssprop --wcs-com-nav-item-arrow-icon-font-size - Arrow size
 * @cssprop --wcs-com-nav-item-focus-outline-color - Focus outline color
 */
export class ComNavItem {
    onClick() {
        this.wcsClickOnFinalAction.emit();
    }
    render() {
        return (h(Host, { key: '25690c3eecac76d2b29b0aeda90f5e1bf5fd0b22', role: "listitem" }, h("slot", { key: '236d67ffd9fefb7650d2b67cee685852030adb4b' }), h("span", { key: '2ef478d77dc28d169817500145c8d9df525bd439', class: "arrow-container" }, h("span", { key: '0fafbba11e817663ad7e18ff6947cdbd07494f69', "aria-hidden": "true", class: "arrow-icon" }, "\uF107"))));
    }
    static get is() { return "wcs-com-nav-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["./com-nav-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["com-nav-item.css"]
        };
    }
    static get events() {
        return [{
                "method": "wcsClickOnFinalAction",
                "name": "wcsClickOnFinalAction",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when a user click on a final navigation action.\n\nUsed by the com-nav component to close the mobile menu overlay when a user click on a final action."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get listeners() {
        return [{
                "name": "click",
                "method": "onClick",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
