// wcs-core/dist/collection/components/dropdown-header/dropdown-header.js
import { h, Host } from "@stencil/core";
/**
 * You can add `wcs-dropdown-header` before a group of items to describe it
 * @slot <no-name> the slot that contains the header's name
 *
 * @cssprop --wcs-dropdown-header-color - Text color of the header
 * @cssprop --wcs-dropdown-header-padding-vertical - Vertical padding of the header
 * @cssprop --wcs-dropdown-header-padding-horizontal - Horizontal padding of the header
 * @cssprop --wcs-dropdown-header-font-size - Font size of the header
 * @cssprop --wcs-dropdown-header-font-weight - Font weight of the header
 *
 */
export class DropdownHeader {
    render() {
        return (h(Host, { key: 'a9f4cf33ebcc666bd3f703c3d51d17b655f28ed8', slot: "item", tabindex: "-1" }, h("slot", { key: 'ccabe9a8805d48d8b2bec9326b5aaa1b6338f06b' })));
    }
    static get is() { return "wcs-dropdown-header"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["dropdown-header.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["dropdown-header.css"]
        };
    }
}
