// wcs-core/dist/collection/components/footer/footer.js
import { h, Host } from "@stencil/core";
/**
 * The footer component has been designed to leave as much customization as possible to the developer. The footers are often specific to the application developed.
 *
 * @cssprop --wcs-footer-background-color - Background color of the footer
 * @cssprop --wcs-footer-text-color - Text color of the footer
 * @cssprop --wcs-footer-link-text-color - Text color of the links in the footer
 * @cssprop --wcs-footer-link-text-color-hover - Text color of the links in the footer when hovered
 * @cssprop --wcs-footer-link-font-size - Font size of the links in the footer
 * @cssprop --wcs-footer-link-font-weight - Font weight of the links in the footer
 * @cssprop --wcs-footer-link-gap - Gap between the links in the footer
 * @cssprop --wcs-footer-gap - Vertical gap between the content and the two end slots
 * @cssprop --wcs-footer-end-gap - Horizontal gap between the right and left end slots
 * @cssprop --wcs-footer-max-width - Max width of the footer for responsive purposes
 * @cssprop --wcs-footer-padding - Padding all around the content of the footer
 *
 * @slot <no-name> Main container slot
 * @slot end-left Bottom-left part of the footer
 * @slot end-right Bottom-right part of the footer
 */
export class Footer {
    render() {
        return (h(Host, { key: '6fe975235d6de2793e55e8d10860649b5684c093' }, h("footer", { key: '80b0ff721eedab3e4cc00fd2484a638718fc8b13', role: "contentinfo" }, h("div", { key: 'f7d676d79f4d8780736e3a28e3e4b2aa1f0786e1', class: "container" }, h("div", { key: 'e1f9ea823133d972d3eb2bc8e7eaa18558448a73' }, h("slot", { key: 'd7177b9e98ac24ad39008669de01899e71c27079' })), h("div", { key: '849ceafdd459f24b672e1f68edf86ffd4fcce733', class: "end" }, h("div", { key: '7077861e872bbb50e7e5de0c968fe2222e9bf7d7', class: "end-left" }, h("slot", { key: 'dded653e62bf683e0514c51b06e55ccb44f70c54', name: "end-left" })), h("div", { key: 'ec25d2a5f84cb477eaa6a63de4df6ce6e9db07c5', class: "end-right" }, h("slot", { key: '4201096294b44a29f74112c2acaa2163f9ecfcb9', name: "end-right" })))))));
    }
    static get is() { return "wcs-footer"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["footer.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["footer.css"]
        };
    }
}
