// wcs-core/dist/collection/components/badge/badge.js
import { h } from "@stencil/core";
/**
 * The badge component is a small label, generally appearing inside or in proximity to another larger interface component,
 * representing a status, property, or some other metadata.
 *
 * @cssprop --wcs-badge-font-weight - Define the font weight of the badge
 * @cssprop --wcs-badge-height-l - Large height of the badge
 * @cssprop --wcs-badge-font-size-l - Large font size of the badge
 * @cssprop --wcs-badge-height-m - Medium height of the badge
 * @cssprop --wcs-badge-font-size-m - Font size of the badge
 * @cssprop --wcs-badge-height-s - Small height of the badge
 * @cssprop --wcs-badge-font-size-s - Small font size of the badge
 * @cssprop --wcs-badge-border-radius-default - Default border radius of the badge
 * @cssprop --wcs-badge-border-radius-circular - Circular border radius of the badge
 *
 * @cssprop --wcs-badge-background-color - Background color of the badge
 * @cssprop --wcs-badge-background-color-lighter - Background color of the lighter badge
 * @cssprop --wcs-badge-color - Text color of the badge
 * @cssprop --wcs-badge-color-lighter - Text color of the lighter badge
 * @cssporp --wcs-badge-gap - The gap inside the badge
 *
 * @cssprop --wcs-badge-padding-horizontal-l - Horizontal padding of the size L badge
 * @cssprop --wcs-badge-padding-horizontal-m - Horizontal padding of the size M badge
 * @cssprop --wcs-badge-padding-horizontal-s - Horizontal padding of the size S badge
 */
export class Badge {
    constructor() {
        this.shape = 'normal';
        this.color = 'initial';
        this.size = 'm';
    }
    render() {
        return (h("slot", { key: '6d9646b28344f4d1f05a3efaf798a6273252a8bf' }));
    }
    static get is() { return "wcs-badge"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["badge.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["badge.css"]
        };
    }
    static get properties() {
        return {
            "shape": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "BadgeShape",
                    "resolved": "\"normal\" | \"rounded\"",
                    "references": {
                        "BadgeShape": {
                            "location": "import",
                            "path": "./badge-interface",
                            "id": "src/components/badge/badge-interface.ts::BadgeShape"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Define the shape of the badge"
                },
                "attribute": "shape",
                "reflect": false,
                "defaultValue": "'normal'"
            },
            "color": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "BadgeColor",
                    "resolved": "\"initial\" | \"lighter\"",
                    "references": {
                        "BadgeColor": {
                            "location": "import",
                            "path": "./badge-interface",
                            "id": "src/components/badge/badge-interface.ts::BadgeColor"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Allows you to change the color of the badge to make it less bright (based on the color chosen by the CSS class)."
                },
                "attribute": "color",
                "reflect": false,
                "defaultValue": "'initial'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "BadgeSize",
                    "resolved": "\"l\" | \"m\" | \"s\"",
                    "references": {
                        "BadgeSize": {
                            "location": "import",
                            "path": "./badge-interface",
                            "id": "src/components/badge/badge-interface.ts::BadgeSize"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Specify the size of the badge."
                },
                "attribute": "size",
                "reflect": true,
                "defaultValue": "'m'"
            }
        };
    }
}
