// wcs-core/dist/collection/components/action-bar/action-bar.js
import { h, Host } from "@stencil/core";
/**
 * The action-bar component is a group of elements appearing across the top of all pages on a business application.
 *
 * @cssprop --wcs-action-bar-background-color - Background color of the action bar
 * @cssprop --wcs-action-bar-gap - Gap between the title and the tabs
 * @cssprop --wcs-action-bar-gap-actions - Gap between the title and the actions
 * @cssprop --wcs-action-bar-min-height - Minimum height of the action bar
 * @cssprop --wcs-action-bar-padding-horizontal - Horizontal padding of the action bar
 * @cssprop --wcs-action-bar-padding-vertical - Vertical padding of the action bar
 * @cssprop --wcs-action-bar-tabs-margin-top - Margin top of the tabs when slotted
 * @cssprop --wcs-action-bar-title-font-size - Font size of the title
 * @cssprop --wcs-action-bar-title-font-weight - Font weight of the title
 * @cssprop --wcs-action-bar-title-color - Color of the title
 * @cssprop --wcs-tabs-gutter-border-width - Border width of the gutter
 * @cssprop --wcs-tabs-gutter-background-color - Background color of the gutter
 */
export class ActionBar {
    constructor() {
        this.gutter = undefined;
        this.hasTabs = false;
    }
    componentWillLoad() {
        this.hasTabs = !!this.el.querySelector('[slot="tabs"]');
    }
    render() {
        return (h(Host, { key: '4312c81132be4f4db1205a5f0bfae699c26e6436' }, h("div", { key: '746da3288b2092209451cf9014d5b5374bc36d70', class: "title-actions", "data-has-tabs": this.hasTabs }, h("h1", { key: 'e143ac0afe1fe76f2760859ba700d432d9d5756d' }, h("slot", { key: 'f5b7269692fd481a0d9fbc0e873580adfe9d6cc1' })), h("div", { key: '0ef5d4aca6cc88c96d541cd391e04a31ead116ee', class: "actions" }, h("slot", { key: '05364550979377548c2e4c97a21f3b52b27b457f', name: "actions" }))), h("div", { key: '613a7c98d9e9777f900e634500e541860a242fbc', class: "tabs-container" }, h("slot", { key: '35b753163a40381873e8fb07eb13fef58c0bcda6', name: "tabs" }))));
    }
    static get is() { return "wcs-action-bar"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["action-bar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["action-bar.css"]
        };
    }
    static get properties() {
        return {
            "gutter": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines if the action bar should have a border at the bottom.\nYou should not use this property if a gutter is already present on tabs"
                },
                "attribute": "gutter",
                "reflect": true
            }
        };
    }
    static get states() {
        return {
            "hasTabs": {}
        };
    }
    static get elementRef() { return "el"; }
}
