// wcs-core/dist/collection/components/grid-custom-cell/grid-custom-cell.js
import { h, Host } from "@stencil/core";
/**
 * The grid custom cell is a subcomponent of `wcs-grid`.
 */
export class GridCustomCell {
    constructor() {
        this.columnId = undefined;
        this.rowId = undefined;
    }
    render() {
        return (h(Host, { key: '3353e4de7a3eeb4d903a63d5adf257ed3427beb8', slot: this.columnId + '-' + this.rowId }, h("slot", { key: 'cc3297d327379b74a385c4909964b8371da68a97' })));
    }
    static get is() { return "wcs-grid-custom-cell"; }
    static get encapsulation() { return "shadow"; }
    static get properties() {
        return {
            "columnId": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "ID of the column for which to render the cell"
                },
                "attribute": "column-id",
                "reflect": false
            },
            "rowId": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Key value of the object rendered for the cell's row"
                },
                "attribute": "row-id",
                "reflect": false
            }
        };
    }
}
